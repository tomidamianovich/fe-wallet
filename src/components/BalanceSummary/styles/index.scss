.fe-wallet__balance {
  display: flex;
  flex-wrap: nowrap;
  justify-content: space-between;
  align-content: center;
  font-weight: 600;
  align-items: flex-end;

  &__symbol {
    margin-right: 0.125rem;
  }
  
  &__amount-desc {
    margin-right: 0.0625rem;
  }
}