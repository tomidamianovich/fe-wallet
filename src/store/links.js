const navLinks = [
  {
    name: 'Panel',
    to: 'panel',
    icon: 'panel'
  },
  {
    name: 'Billetera',
    to: 'wallet',
    icon: 'wallet'
  },
  {
    name: 'Cotizaciones',
    to: 'market-rates',
    icon: 'market-rates'
  },
  {
    name: 'Cr√©ditos',
    to: 'credit',
    icon: 'credit'
  },
  {
    name: 'Exchange',
    to: 'exchange',
    icon: 'exchange'
  },
  {
    name: 'Launchpad',
    to: 'launchpad',
    icon: 'launchpad'
  }
];

export default navLinks;