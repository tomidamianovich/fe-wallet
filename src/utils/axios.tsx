import { RateType, CurrencyType } from '../utils/type';
const axios = require('axios').default;

type responseType = {
  data: RateType[] | CurrencyType[]
}
  
const requestHandler = async (url:string): Promise<any> => 
  new Promise((resolve, reject) => 
    axios.get(url)
    .then((response: responseType) => resolve(response?.data))
    .catch((err:any) => reject(err))
  );

export {
  requestHandler,
}