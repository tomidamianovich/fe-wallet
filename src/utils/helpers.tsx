const handleFormatValues = (value:number) => value.toFixed(2);

export {
  handleFormatValues
}