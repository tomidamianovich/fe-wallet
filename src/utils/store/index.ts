import currencies from './currencies';
import cryptoCurrencies from './cryptoCurrencies';
import fiatCurrencies from './fiatCurrencies';
import balances from './balances';
import rates from './rates';

export {
  currencies,
  fiatCurrencies,
  cryptoCurrencies,
  rates,
  balances
}