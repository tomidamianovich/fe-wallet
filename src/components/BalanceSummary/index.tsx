import React, { useState } from 'react';
import { useEffect } from 'react';
import { BalanceType, CurrencyType } from '../../utils/type';
import './styles/index.scss';
import Dropdown from '../Dropdown'

type Props = {
  balances: BalanceType[],
  currencies: CurrencyType[]
};

const BalanceSummary: React.FC<Props> = ({
  balances,
  currencies
}) => {
  const [currentBalance, setCurrentBalance] = useState<BalanceType>(balances[0]);
  const [currentCurrencyIndex, setCurrentCurrencyIndex] = useState<number>(0);

  useEffect(()=> {
    const currencyIndex = currencies.findIndex(currency => currency.ticker === currentBalance.ticker);
    currencyIndex >= 0 && setCurrentCurrencyIndex(currencyIndex);
  }, [currentBalance, currencies]);


  const handlerDropdown = (e: React.ChangeEvent<HTMLInputElement>) => {
    const currencyIndex = parseInt(e.target.value);
    const balance = balances.find(balance => balance.ticker === currencies[currencyIndex].ticker);
    if (!balance) return
    setCurrentCurrencyIndex(currencyIndex);
    setCurrentBalance(balance);
  };

  const getFormattedAmount = (amount: number, decimals: number):string => {
    return (Math.round(amount * 100) / 100).toFixed(decimals);
  }

  return(
    <div className="fe-wallet__balance">
      <span className="fe-wallet__balance__symbol">
        { currencies[currentCurrencyIndex]?.symbol }
      </span>
      <span className="fe-wallet__balance__amount-desc">
        { getFormattedAmount(currentBalance?.amount, currencies[currentCurrencyIndex]?.decimals ) }
      </span>
      <Dropdown
        name="balance" 
        handler={handlerDropdown} 
        options={currencies} 
        selectedIndex={currentCurrencyIndex}
        ariaLabel={"Seleccionar moneda para ver balance"}
        className="fe-wallet__dropdown__balance"
      />
    </div>  
)};

export default BalanceSummary;