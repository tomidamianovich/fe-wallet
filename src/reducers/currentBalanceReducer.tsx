import * as actionTypes from "../actionTypes/BalanceType"
import { BalanceType, CurrentBalanceAction } from '../utils/type'

const initialState:BalanceType = {
  ticker: "",
  amount: 0,
  account_limit: 0
};

const reducer = (
  state: BalanceType = initialState,
  action: CurrentBalanceAction
): BalanceType => {
  switch (action.type) {
    case actionTypes.SET_CURRENT_BALANCE:
      const balances:BalanceType = action.payload
      return balances;
  }
  return state
}

export default reducer
