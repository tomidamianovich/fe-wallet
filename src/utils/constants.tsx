import { transaction, transactionTitle, transactionStatus } from '../utils/type';

const ENDPOINTS = {
  CURRENCIES: 'http://localhost:3001/currencies',
  RATES: 'http://localhost:3001/rates',
  BALANCES: 'http://localhost:3001/balance'
}

const PANEL = {
  HEADING: "Mi billetera",
  SUBHEADING: "En tu billetera vas a poder almacenar todas las criptomonedas que compres en ripio.",
}

const TRANSACTIONS:{
  HEADING: string,
  HEADER: string[],
  TYPES: transaction,
  TYPES_TITLES: transactionTitle,
  STATUS: transactionStatus
} = {
  HEADING: "Mis movimientos",
  HEADER: [
    "Fecha",
    "Tipo de Operación",
    "Método",
    "Estado",
    "Comisión",
    "Monto",
  ],
  TYPES: {
    DEPOSIT: "DEPOSIT",
    WITHDRAWAL: "WITHDRAWAL",
    SEND: "SEND",
    RECEIVE: "RECEIVE",
    BUY: "BUY",
    SWAP: "SWAP"
  },
  TYPES_TITLES: {
    DEPOSIT: "Ingreso",
    WITHDRAWAL: "Retiro",
    SEND: "Envio",
    RECEIVE: "Recibo",
    BUY: "Compra",
    SWAP: "Swap"
  },
  STATUS: {
    PENDING: "PENDING",
    FAILURE: "FAILURE",
    SUCCESS: "SUCCESS"
  }
}

const DEFAULT_BALANCE = {
  ticker: "",
  amount: 0,
  account_limit: 0,
  name: "",
  symbol: "",
  decimals: 2,
  type: "",
  actions: [{
    action: "",
    enabled: true
  }],
  url_images: {
    image_png: "",
    image_svg: "",
  },
  color: "",
  categories: [],
  is_favorite: true,
  sell_rate: ""
}

const COMISIONS = {
  BUY: 0.003,
  DEPOSIT: 0,
  WITHDRAWAL: 0.0004,
  SEND: 0,
  RECEIVE: 0,
  SWAP: 0.00004
}

type Link = {
  name: string;
  to: string;
  icon: string;
};

const NAV_LINKS:Link[] = [
  {
    name: 'Panel',
    to: 'panel',
    icon: 'panel'
  },
  {
    name: 'Billetera',
    to: 'wallet',
    icon: 'wallet'
  },
  {
    name: 'Cotizaciones',
    to: 'market-rates',
    icon: 'market-rates'
  },
  {
    name: 'Créditos',
    to: 'credit',
    icon: 'credit'
  },
  {
    name: 'Exchange',
    to: 'exchange',
    icon: 'exchange'
  },
  {
    name: 'Launchpad',
    to: 'launchpad',
    icon: 'launchpad'
  }
];

const WORDINGS = {
  BALANCE: {
    DETAIL: {
      INPUT: {
        LABELS: {
          TRANSACTION: "Seleccione el tipo de operacion a realizar",
          AMOUNT: "Ingrese cantidad",
          ARIA_LABEL : "Monto de la transacción"
        },
        MESSAGES: {
          INSUFFICIENT_AMOUNT: "Dinero insuficiente",
          AMOUNT_TO_DEBIT: "Monto a debitar",
          AMOUNT_NEEDED: "Monto necesario",
          ACTUAL_BALANCE: "Monto actual"
        }
      },
      ACTIONS: {
        CANCEL: {
          LABEL: "Cancelar",
          ARIA_LABEL: "Cancelar transacción",
        },
        CONFIRM: {
          LABEL: "Confirmar",
          ARIA_LABEL: "Confirmar transacción",
        }
      },
      CARD: {
        ACCOUNT: "Cuenta"
      }
    },
    SUMMARY: {
      DROPDOWN: {
        ARIA_LABEL: "Seleccionar moneda para ver balance"
      }
    }
  },
  CONVERTER: {
    FROM: {
      DROPDOWN: {
        ARIA_LABEL: "Seleccionar crypto origen",
        NAME: "from"
      },
      INPUT: {
        ARIA_LABEL: "Cripto a convertir",
        NAME: "from_amount"
      }
    },
    TO: {
      DROPDOWN: {
        ARIA_LABEL: "Seleccionar crypto destino",
        NAME: "to"
      },
      INPUT: {
        ARIA_LABEL: "Cripto convertida",
        NAME: "to_amount"
      }
    },
    BUTTON: {
      ARIA_LABEL: "Convertir",
      LABEL: "Convertir",
    }
  },
  CURRENCIES: {
    TYPES: {
      FIAT: "FIAT",
      CRYPTO: "CRYPTO"
    }
  },
  MAINTENANCE: {
    TITLE: "Pagina en Mantenimiento",
    SUBTITLE: {
      BEGINNING: "Oops! Estamos trabajando en la pagina ",
      ENDING: ". Ingrese nuevamente a esta seccion más tarde."
    }
  },
  NAVIGATION: {
    NAV_ITEM: {
      NAME: "Salir"
    }
  },
  NOT_FOUND: "Pagina no encontrada.",
  ERROR_FOUND: "Oops. Ocurrio un error",
  PAGES: {
    WALLET: "wallet",
    CREDIT: "credit",
    EXCHANGE: "exchange",
    LAUNCHPAD: "launchpad",
    MARKET_RATES: "market-rates",
    PANEL: "panel",
  }
}

export {
  ENDPOINTS,
  PANEL,
  TRANSACTIONS,
  DEFAULT_BALANCE,
  COMISIONS,
  NAV_LINKS,
  WORDINGS
}