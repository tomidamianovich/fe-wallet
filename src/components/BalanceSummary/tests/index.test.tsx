import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import BalanceSummary from '../index';
import { balances, fiatCurrencies, currencies } from '../../../utils/store';
import currentBalanceReducer from "../../../reducers/currentBalanceReducer"
import { Provider } from 'react-redux';
import { createStore } from "redux";

const store = createStore(currentBalanceReducer);

test('renders BalanceSummary with balances', () => {
  const { asFragment } = render(
    <Provider store={store}>
      <BalanceSummary 
        balances={balances}
        currencies={currencies}
      />
    </Provider>
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders BalanceSummary with balances and switching between currencies', async () => {
  render(
    <Provider store={store}>
      <BalanceSummary 
        balances={balances}
        currencies={fiatCurrencies}
      />
    </Provider>
  );
  
  const currencySelector = screen.getByRole('combobox', { name: "Seleccionar moneda para ver balance" });
  expect(currencySelector).toBeInTheDocument();
  expect(currencySelector).toHaveValue("0");
  expect(currencySelector).toHaveDisplayValue("ARS");

  const currencySelectorArs = screen.getByRole('option', { name: "ARS" });
  const currencySelectorUsd = screen.getByRole('option', { name: "USD" });
  expect(currencySelectorArs).toHaveValue("0");
  expect(currencySelectorUsd).toHaveValue("1");

  expect(currencySelector).toHaveValue("0");

  fireEvent.change(currencySelector, { target: { value: "1" }});
  await waitFor(() => {
    expect(currencySelector).toHaveValue("1");
    expect(currencySelector).toHaveDisplayValue("USD");
    expect(screen.findByDisplayValue("U$S")).toBeTruthy();
  });

  fireEvent.change(currencySelector, { target: { value: "0" }});
  await waitFor(() => {
    expect(currencySelector).toHaveValue("0");
    expect(currencySelector).toHaveDisplayValue("ARS");
    expect(screen.findByDisplayValue("$")).toBeTruthy();
  });
  
});

