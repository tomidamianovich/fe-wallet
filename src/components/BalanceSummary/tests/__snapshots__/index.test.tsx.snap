// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders BalanceSummary with balances 1`] = `
<DocumentFragment>
  <div
    class="fe-wallet__balance"
  >
    <span
      class="fe-wallet__balance__symbol"
    >
      $
    </span>
    <span
      class="fe-wallet__balance__amount-desc"
    >
      NaN
    </span>
    <select
      aria-label="Seleccionar moneda para ver balance"
      class="fe-wallet__dropdown fe-wallet__dropdown__balance"
      id="balance"
      name="balance"
    >
      <option
        value="0"
      >
        ARS
      </option>
      <option
        value="1"
      >
        BTC
      </option>
      <option
        value="2"
      >
        ETH
      </option>
      <option
        value="3"
      >
        USDC
      </option>
      <option
        value="4"
      >
        LTC
      </option>
      <option
        value="5"
      >
        DAI
      </option>
      <option
        value="6"
      >
        AXS
      </option>
      <option
        value="7"
      >
        BAT
      </option>
      <option
        value="8"
      >
        SOL
      </option>
      <option
        value="9"
      >
        LINK
      </option>
      <option
        value="10"
      >
        DOT
      </option>
      <option
        value="11"
      >
        MANA
      </option>
      <option
        value="12"
      >
        RPC
      </option>
      <option
        value="13"
      >
        SLP
      </option>
      <option
        value="14"
      >
        ADA
      </option>
      <option
        value="15"
      >
        UNI
      </option>
      <option
        value="16"
      >
        CHZ
      </option>
      <option
        value="17"
      >
        XRP
      </option>
      <option
        value="18"
      >
        USDT
      </option>
      <option
        value="19"
      >
        DOGE
      </option>
      <option
        value="20"
      >
        AAVE
      </option>
      <option
        value="21"
      >
        UBI
      </option>
    </select>
  </div>
</DocumentFragment>
`;
